ts/defs.bzl

```diff
@@ -313,19 +325,43 @@ def ts_project(
     typings_out_dir = declaration_dir if declaration_dir else out_dir
     tsbuildinfo_path = ts_build_info_file if ts_build_info_file else name + ".tsbuildinfo"
 
-    tsc_typings_outs = _lib.calculate_typings_outs(srcs, typings_out_dir, root_dir, declaration, composite, allow_js, no_emit)
-    tsc_typing_maps_outs = _lib.calculate_typing_maps_outs(srcs, typings_out_dir, root_dir, declaration_map, allow_js, no_emit)
+    # Target names for tsc, dts+js transpilers
+    tsc_target_name = name
+    declarations_target_name = None
+    transpile_target_name = None
+
+    # typing outputs
+    tsc_typings_outs = []
+    tsc_typing_maps_outs = []
+    if not declaration_transpiler or declaration_transpiler == "tsc":
+        tsc_typings_outs = _lib.calculate_typings_outs(srcs, typings_out_dir, root_dir, declaration, composite, allow_js, no_emit)
+        tsc_typing_maps_outs = _lib.calculate_typing_maps_outs(srcs, typings_out_dir, root_dir, declaration_map, allow_js, no_emit)
```

Comment:
I think these are not actually used since they are passed to `ts_project_rule`, however `ts_project_rule` never reads from `ctx.attr.typings_outs` nor `ctx.attr.typing_maps_outs`. It recalculates the outs itself.

Reply:
It's done in the macro here just to predeclare all possible, so the outputs can be referenced as labels.

Then within the rule they are ignored and recalculated like you mentioned, then things like targets are expanded to a full list of files which could not be pre-calculated.